# Project 2, Part 5, Cleansing customer data

University of California, Berkeley

Master of Information and Data Science (MIDS) program

w205 - Fundamentals of Data Engineering

Student: Landon Morin

Year: 2022

Semester: Spring

Section: 9


# Included Modules and Packages

Code cell containing your includes for modules and packages

In [1]:
import math
import numpy as np
import pandas as pd
import time

import psycopg2

import json

import csv

from datetime import datetime as dt

from IPython.display import display, HTML

from jellyfish import soundex, levenshtein_distance

from fuzzywuzzy import fuzz

from fuzzywuzzy import process as fuzz_process


# Supporting code

Code cells containing any supporting code, such as connecting to the database, any functions, etc.  

Remember you can freely use any code from the labs. You do not need to cite code from the labs.

In [2]:
connection = psycopg2.connect(
    user = "postgres",
    password = "ucb",
    host = "postgres",
    port = "5432",
    database = "postgres"
)

In [3]:
cursor = connection.cursor()

In [4]:
#
# function to run a select query and return rows in a pandas dataframe
# pandas puts all numeric values from postgres to float
# if it will fit in an integer, change it to integer
#

def my_select_query_pandas(query, rollback_before_flag, rollback_after_flag):
    "function to run a select query and return rows in a pandas dataframe"
    
    if rollback_before_flag:
        connection.rollback()
    
    df = pd.read_sql_query(query, connection)
    
    if rollback_after_flag:
        connection.rollback()
    
    # fix the float columns that really should be integers
    
    for column in df:
    
        if df[column].dtype == "float64":

            fraction_flag = False

            for value in df[column].values:
                
                if not np.isnan(value):
                    if value - math.floor(value) != 0:
                        fraction_flag = True

            if not fraction_flag:
                df[column] = df[column].astype('Int64')
    
    return(df)
    

# 2.5.1 Validate the city, state, and zip for stage_1_peak_customers against the zip_codes table

AGM does not want to give its customer list to 3rd party sales channels, including Peak Delivery.  For that reason, we can expect some variation in customer first and last names, and in the street.  However, the city, state, and zip should be validated by Peak's system, so we do not anticipate any issues.

Write a query that demonstrates that the city, state, and zip are valid for all records.  Like we did in 2.4, it's usually best to write a query that return errors.  In our case the query should not return anything.

Sort by stage_id

Pattern your code after the examples in the labs.  You may use as many code cells as you need.

In [5]:
rollback_before_flag = True
rollback_after_flag = True

query  = """

select *
from stage_1_peak_customers
where (city, state, zip) not in (select city, state, zip from zip_codes)
order by stage_id


"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

Empty DataFrame
Columns: [stage_id, sale_id, customer_id, first_name, last_name, street, city, state, zip]
Index: []

# 2.5.2 Find all customer records in stage_1_peak_customers where any of first_name, last_name, and/or street do not match a customer in the customers table

AGM does not want to give its customer list to 3rd party sales channels, including Peak Delivery.  For that reason, we can expect some variation in customer first and last names, and in the street.  

Write a query that returns all customer records in state_1_peak_customers where any of the first_name, last_name, and/or street do not match a customer in the customers table

Sort by stage_id

Pattern your code after the examples in the labs.  You may use as many code cells as you need.

In [6]:
rollback_before_flag = True
rollback_after_flag = True

query  = """

select *
from stage_1_peak_customers
where (first_name, last_name, street) not in (select first_name, last_name, street from customers)
order by stage_id


"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

stage_id     sale_id customer_id first_name last_name  \
0         10  5763728768     3729016      Hyrum  Knuckles   
1         20  5763728877     3728936    Roseann    Coyish   
2         24  5763728428     3729287       Hali    Ducker   
3         26  5763728393     3728674   Melantha   Golborn   
4         36  5763729212     3729191      Eleni    Jansen   
5         40  5763729129     3728856      Clyve   Humonds   
6         51  5763728864     3729178   Rutledge  Hellwing   
7         60  5763729313     3728402      Kalli     Kemel   
8         72  5763728980     3729213     Honina   Philson   
9         73  5763728921     3729194      Nicky     Haley   
10        75  5763729280     3729104          I  Odunniom   
11        78  5763728905     3729208    Clemens  McAloren   
12        97  5763728673     3728691       Eula    Dewair   

                    street        city state    zip  
0    86668 Spenser Terrace     Oakland    CA  94618  
1   11707 American Ash Ter      Orinda    CA  94563  
2             8 Orion Pass  El Cerrito    CA  94530  
3   6140 North Field Alley      Orinda    CA  94563  
4          66 Bartelt Hill     Oakland    CA  94607  
5       22 Brent Wood Hill    Berkeley    CA  94709  
6             606 Gulf Plz  El Cerrito    CA  94530  
7     18373 Golf View Pass    Berkeley    CA  94702  
8       28 Clarendon Plaza    Berkeley    CA  94702  
9       88424 Warrior Lane     Oakland    CA  94602  
10         4243 Sugar Park    Berkeley    CA  94702  
11      12532 Faragut Lane    Berkeley    CA  94704  
12       3 Crownhardt Road      Albany    CA  94706

# 2.5.3 Find the percentage of Peak's customer records that do not match to AGM's customers table

Write a query to find the percentage of Peak's customer records that do not match AGM's.  The percentage can be found by taking the number of customer records in stage_1_peak_customers that do not match and dividing by the number of customers records in stage_1_peak_customers and multiplying by 100.

Show the total number of Peak customer records in stage_1_peak_customers, the number that match to customers, the number that do not match to customers, and the percentage that do not match.

Show the percentage rounded to the nearest tenth.  It is not necessary to include a percent sign.

Pattern your code after the examples in the labs.  You may use as many code cells as you need.

In [7]:
rollback_before_flag = True
rollback_after_flag = True

query  = """

with a as (select count(customer_id) as non_matching_customers_total
           from stage_1_peak_customers
           where (first_name, last_name, street) not in (select first_name, last_name, street from customers)), 
           
b as (select count(customer_id) as total_customers
      from stage_1_peak_customers)
           
select total_customers, non_matching_customers_total, 
       b.total_customers - a.non_matching_customers_total as total_matching_customers,
       round(a.non_matching_customers_total::numeric/b.total_customers::numeric*100,1) as peak_unmatched_customer_percentage
from a, b



"""

my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

total_customers  non_matching_customers_total  total_matching_customers  \
0               97                            13                        84   

   peak_unmatched_customer_percentage  
0                                13.4

# 2.5.4 Executive summary on customer data

Write an executive summary on the customer data.  

The summary should be the equivalent to 3/4 to 1 page using standard fonts, spacing, and margins. 

As stated in the scenario, like most companies, AGM does not want to give out its customer list to 3rd party sales channels.  The downside is, as we have seen, that customer first names, last names, and street addresses will have some variations and not be exact matches.

The executives would like your recommendation of one of the following:
* Continue to withhold the customer data from 3rd party sales channels
* Give customer data to 3rd party sales channels

Recommend exactly one of these.

Support you recommendation with an explanation based on what you have seen from this preliminary data load.

You are not required to write any queries nor create any data visualizations.  However, you may want to include some to enhance and add quality to your submission.  Submissions with these tend to be higher quality, although, not always.

You may use any number of code cells and/or markdown cells. 

You may alternate between code cells and markdown cells.  That is perfectly fine.  It is understood that before we present it, an editor would pull out the text, results of queries, and data visualizations.

In [8]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select c.customer_id,
       pc.first_name as peak_first,
       pc.last_name as peak_last,
       c.first_name as agm_first, 
       c.last_name as agm_last,
       pc.street as peak_street, 
       c.street as agm_street
from stage_1_peak_customers as pc
    join customers as c
         on pc.zip = c.zip
where (pc.first_name, pc.last_name, pc.street) not in (select c.first_name, c.last_name, c.street from customers)
order by stage_id



"""
cursor.execute(query)

connection.rollback()
    
rows = cursor.fetchall()


In [9]:

start = time.time()

for row in rows:
    if (levenshtein_distance(row[1], row[3]) < 4 and 
    levenshtein_distance(row[2], row[4]) < 4 and 
    levenshtein_distance(row[5], row[6]) < 4):
        print("---------------------------------------------------------")
        print('Possible AGM Customer id:', row[0])
        print('First Name PEAK:', row[1], '\nFirst Name AGM:', row[3])
        print('Last Name PEAK:' , row[2], '\nLast Name AGM:', row[4])
        print('Address PEAK:' , row[5], '\nAddress AGM:', row[6])
        print("Levenshtein Distance First Name:", levenshtein_distance(row[1], row[3]))
        print("Levenshtein Distance Last Name:", levenshtein_distance(row[2], row[4]))
        print("Levenshtein Distance Street:", levenshtein_distance(row[5], row[6]))


end = time.time()
print('time:', end - start)

---------------------------------------------------------
Possible AGM Customer id: 281
First Name PEAK: Hyrum 
First Name AGM: Hiram
Last Name PEAK: Knuckles 
Last Name AGM: Knuckles
Address PEAK: 86668 Spenser Terrace 
Address AGM: 86668 Spenser Terrace
Levenshtein Distance First Name: 2
Levenshtein Distance Last Name: 0
Levenshtein Distance Street: 0
---------------------------------------------------------
Possible AGM Customer id: 3163
First Name PEAK: Hali 
First Name AGM: Halli
Last Name PEAK: Ducker 
Last Name AGM: Ducker
Address PEAK: 8 Orion Pass 
Address AGM: 8 Orin Pass
Levenshtein Distance First Name: 1
Levenshtein Distance Last Name: 0
Levenshtein Distance Street: 1
---------------------------------------------------------
Possible AGM Customer id: 2271
First Name PEAK: Melantha 
First Name AGM: Melantha
Last Name PEAK: Golborn 
Last Name AGM: Golborn
Address PEAK: 6140 North Field Alley 
Address AGM: 6140 Northfield Alley
Levenshtein Distance First Name: 0
Levenshtein Di

In [10]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select *
from stage_1_peak_customers
where (first_name, last_name, street) not in (select first_name, last_name, street from customers)
order by stage_id



"""

nonmatch = my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

In [11]:
rollback_before_flag = True
rollback_after_flag = True

query = """

select *
from customers
order by customer_id



"""

agm = my_select_query_pandas(query, rollback_before_flag, rollback_after_flag)

In [12]:
dfa = pd.DataFrame({'customer':[]})
dfa['customer'] = agm['first_name']+', '+agm['last_name']+", "+agm['street']

In [13]:
dfp = pd.DataFrame({'customer':[]})
dfp['customer'] = nonmatch['first_name']+', '+nonmatch['last_name']+", "+nonmatch['street']

In [14]:
dfp

customer
0      Hyrum, Knuckles, 86668 Spenser Terrace
1     Roseann, Coyish, 11707 American Ash Ter
2                  Hali, Ducker, 8 Orion Pass
3   Melantha, Golborn, 6140 North Field Alley
4              Eleni, Jansen, 66 Bartelt Hill
5          Clyve, Humonds, 22 Brent Wood Hill
6            Rutledge, Hellwing, 606 Gulf Plz
7          Kalli, Kemel, 18373 Golf View Pass
8         Honina, Philson, 28 Clarendon Plaza
9            Nicky, Haley, 88424 Warrior Lane
10               I, Odunniom, 4243 Sugar Park
11      Clemens, McAloren, 12532 Faragut Lane
12            Eula, Dewair, 3 Crownhardt Road

In [15]:
peak_unmatched_list = dfp['customer'].to_list()
peak_unmatched_list

['Hyrum, Knuckles, 86668 Spenser Terrace',
 'Roseann, Coyish, 11707 American Ash Ter',
 'Hali, Ducker, 8 Orion Pass',
 'Melantha, Golborn, 6140 North Field Alley',
 'Eleni, Jansen, 66 Bartelt Hill',
 'Clyve, Humonds, 22 Brent Wood Hill',
 'Rutledge, Hellwing, 606 Gulf Plz',
 'Kalli, Kemel, 18373 Golf View Pass',
 'Honina, Philson, 28 Clarendon Plaza',
 'Nicky, Haley, 88424 Warrior Lane',
 'I, Odunniom, 4243 Sugar Park',
 'Clemens, McAloren, 12532 Faragut Lane',
 'Eula, Dewair, 3 Crownhardt Road']

In [16]:
agm_customer_list = dfa['customer'].to_list()
agm_customer_list

['Robb, Weaving, 5 Ramsey Place',
 'Robby, Belliard, 6 Londonderry Plaza',
 'Sadella, Caudrelier, 548 Mcguire Parkway',
 'Holmes, Shimmings, 99 Kennedy Court',
 'Beverley, Gubbin, 51 Mcbride Drive',
 'Pavia, Millery, 463 Columbus Pass',
 'Engracia, Jeanon, 24 Hanover Court',
 'Melinda, Stodd, 568 Nova Way',
 'Dinnie, Leek, 38 Golf Alley',
 'Caitrin, Filipowicz, 61 Michigan Drive',
 'Sybille, Vipan, 75841 2nd Hill',
 'Genevieve, Roughey, 129 Katie Street',
 'Artie, Brandes, 3744 Troy Circle',
 'Matilda, Keynes, 73 Grayhawk Park',
 'Carmina, Wandrey, 96472 Stephen Alley',
 'Tailor, Ades, 50 Union Center',
 'Davidson, Stapleton, 5 2nd Way',
 'Julio, Vearncomb, 600 Crownhardt Drive',
 'Alysa, Laight, 8 Sunfield Hill',
 'Ragnar, Keeping, 3 Iowa Avenue',
 'Kara, Faulder, 2928 Aberg Drive',
 'Edie, Wholesworth, 9393 Crescent Oaks Parkway',
 'Laurene, Righy, 85651 Daystar Pass',
 'Elvyn, Rucklesse, 38 Becker Center',
 'Theodore, Knaggs, 25 Katie Center',
 'Putnam, Heaford, 471 Knutson Place',


In [17]:
start = time.time()
match_list = []
for cust in peak_unmatched_list:
    match_list.append(fuzz_process.extract(cust, agm_customer_list, limit = 1))
end = time.time()
print('time:', end - start)


    

time: 16.840769052505493


In [18]:
match_list

[[('Hiram, Knuckles, 86668 Spenser Terrace', 95)],
 [('Roseann, Coyish, 11707 American Ash Terrace', 95)],
 [('Halli, Ducker, 8 Orin Pass', 96)],
 [('Melantha, Golborn, 6140 Northfield Alley', 99)],
 [('Eleni, Jansen, 6 Bartelt Hill', 98)],
 [('Clyve, Hammonds, 22 Brentwood Hill', 94)],
 [('Rutledge, Hellwing, 606 Golf Plaza', 94)],
 [('Kalli, Kemell, 18373 Golf View Pass', 99)],
 [('Junina, Philson, 28 Clarendon Plaza', 94)],
 [('Nickey, Haley, 88424 Warrior Lane', 98)],
 [("Israel, O'Dunniom, 4243 Sugar Park", 90)],
 [('Klemens, MacAloren, 12532 Farragut Lane', 95)],
 [('Eula, Deware, 3 Crownhardt Road', 97)]]

# Executive Summary

As our ELT and preliminary analytics have demonstrated, there can be data validation issues when using third party data. Specifically, in one day of trial-deliveries, we have found 13 customer entries in Peak's system that have not matched our customer database. This means that 13% of our customer data is inaccurate useless unless corrected. However, in our first day our analytics team have found multiple avenues for cleansing and validating Peak's customer data. Preliminary analytics for our first day of the delivery trial support our team's recommendation that we do not share our customer data with Peak or any other third party vendor. 

By employing Extract, Load, Transform (ELT) and Continually Integrating/Continually Developing (CI/CD) techniques, our analytics team is able to work with Peak's POS data quickly to resolve inaccuracies in the customer database and perform meaningful analytics. We can guarantee more accurate customer data even with streaming POS data. 

Peak's first day dataset contained 13 entries that did not match with AGM's customer database. However, by creating a script to iterate through AGM's customer entries and Peak's customer entries and to search for minimized Levenshtein Distances, we were able to shrink the error rate from 13.4\% to 2\%. Furthermore, by searching for AGM-Peak customer pairs with the highest Fuzzy Logic scores, we can reduce this error rate to 0. However, there is a time-cost associated with this method. While Fuzzy Logic will guarantee to show us the closest match, filtering by Levenshtein Distances less than 4 gives us satisfactory accuracy and is roughly 41x less time-intensive with this amount of data. At higher data levels, finding matches using Fuzzy Logic will likely become unacceptably time-intensive. Nontheless, we have two methods that will allow us to clean and use inaccurate or unclean data from Peak's database. 


Therefore, we do not recommend sharing our customer data with Peak, since there is no reward for doing so. The risks are many. Firstly, Peak may merge with a larger competitor such as Amazon. If Amazon acquires our customer data, they may be able to outcompete us in our region. Secondly, we lose control of our data, which poses great risks. We cannot guarantee that Peak will treat our customer data with the same care and respect that we do. Finally, Peak may suffer a data breach. This exposes our customers and AGM to unnecessary risks for minimal reward. 


